src/jvm/main/org/jetbrains/kotlinx/lincheck/IdeaPlugin.kt

```diff
@@ -0,0 +1,166 @@
+@file:JvmName("IdeaPluginKt")
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2023 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck
+import org.jetbrains.kotlinx.lincheck.runner.*
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.*
+
+const val MINIMAL_PLUGIN_VERSION = "0.0.1"
+
+// ============== This methods are used by debugger from IDEA plugin to communicate with Lincheck ============== //
+
+/**
+ * Invoked from the strategy [ModelCheckingStrategy] when Lincheck found a bug.
+ * The Debugger creates a breakpoint on this method, so when it's called, debugger receives all the information about the
+ * failed test.
+ * This method is called first, to provide all required information,
+ * then [beforeEvent] method is called on each trace point.
+ *
+ * @param failureType string representation of the failure type.
+ * (`INCORRECT_RESULTS`, `OBSTRUCTION_FREEDOM_VIOLATION`, `UNEXPECTED_EXCEPTION`, `VALIDATION_FAILURE`, `DEADLOCK`).
+ * @param trace failed test trace, where each trace point is represented as a string
+ * (because it's the easiest way to provide some information to the debugger).
+ * @param version current Lincheck version
+ * @param minimalPluginVersion minimal compatible plugin version
+ */
+@Suppress("UNUSED_PARAMETER")
+fun testFailed(
+    failureType: String,
+    trace: Array<String>,
+    version: String?,
+    minimalPluginVersion: String
+) {
+}
+
+/**
+ * Debugger replaces the result of this method to `true` if idea plugin is enabled.
```

Comment:
Doesn't it simply set the system property you are reading?

Reply:
No, we with @zuevmaxim think that setting a system-wide property is less convenient than using an existing approach with an interception of the exact method result.

Reply:
Explain it in the comments then, and please explain why.

Reply:
The main problem with the system property is that we need to add it to all debug processes when the plugin enabled. On the other hand, we have a stable and precise way of communication via breakpoints. Also, it is used in other part, so the obvious choice is to use it here also.

Reply:
Thanks for the clarification. Could you please put it into the code documentation, please?

Reply:
I don't know what to put in code documentation, as we use such a way of communication in this file, while system property is used only for tests